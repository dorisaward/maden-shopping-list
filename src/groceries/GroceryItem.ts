export type GroceryItem = {
  /**
   * Unique identifier for the grocery item
   */
  id: string;

  /**
   * The human readable name for the grocery item
   */
  name: string;

  /**
   * The nummber of the groceries to display on the list
   */
  quantity: number;
};
