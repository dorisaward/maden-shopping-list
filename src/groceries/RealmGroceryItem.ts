import Realm, {ObjectSchema} from 'realm';

export class RealmGroceryItem extends Realm.Object<RealmGroceryItem> {
  _id!: Realm.BSON.ObjectId;
  name!: string;
  quantity!: number;
  static schema: ObjectSchema = {
    name: 'RealmGroceryItem',
    properties: {
      _id: 'objectId',
      name: 'string',
      quantity: 'int',
    },
    primaryKey: '_id',
  };
}
