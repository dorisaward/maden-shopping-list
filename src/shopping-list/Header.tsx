import {Text, View} from 'react-native';
import {styles} from '../styles';
import React from 'react';

export const Header = (): React.JSX.Element => (
  <View style={styles.background}>
    <Text style={styles.headerText}>Mayden Shopping List</Text>
  </View>
);
