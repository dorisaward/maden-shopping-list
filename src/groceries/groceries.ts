import {GroceryItem} from './GroceryItem';

export const groceries: GroceryItem[] = [
  {name: 'broccoli', quantity: 1, id: '0'},
  {name: 'spinach', quantity: 2, id: '1'},
  {name: 'potatoes', quantity: 5, id: '2'},
  {name: 'kidney beans', quantity: 1, id: '3'},
];
