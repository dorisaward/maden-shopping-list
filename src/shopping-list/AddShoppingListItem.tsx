import React, {useCallback, useState} from 'react';
import {GroceryItem} from '../groceries/GroceryItem';
import {View, TextInput, TouchableOpacity, Text, Alert} from 'react-native';
import {styles} from '../styles';
import uuid from 'react-native-uuid';

type Props = {
  addGroceryItem: (groceryItem: GroceryItem) => void;
};

export const AddShoppingListItem = ({
  addGroceryItem,
}: Props): React.JSX.Element => {
  const [newGroceryName, setNewGroceryName] = useState<string | undefined>(
    undefined,
  );
  const [newGroceryQuantity, setNewGroceryQuantity] = useState<number>(0);

  const setQuantity = useCallback(
    (text: string) => {
      try {
        const number = Number(text);
        setNewGroceryQuantity(number);
      } catch (e) {
        console.warn(`Could not cast string ${text} to number`, e);
      }
    },
    [setNewGroceryQuantity],
  );

  const saveGroceryItem = useCallback(() => {
    if (!newGroceryName) {
      Alert.alert('Please make sure your groceries have a name');
      return;
    }
    if (!newGroceryQuantity) {
      Alert.alert(`Please specify how much ${newGroceryName} you want.`);
      return;
    }
    addGroceryItem({
      name: newGroceryName,
      quantity: newGroceryQuantity,
      id: uuid.v4().toString(),
    });
    setNewGroceryQuantity(0);
    setNewGroceryName(undefined);
  }, [newGroceryName, newGroceryQuantity, addGroceryItem]);

  return (
    <View style={styles.shoppingListItem}>
      <TextInput
        placeholder="Name"
        value={newGroceryName}
        style={styles.shoppingListItemInput}
        onChangeText={setNewGroceryName}
      />
      <TextInput
        placeholder="Quantity"
        value={newGroceryQuantity.toString()}
        keyboardType="number-pad"
        style={styles.shoppingListItemInput}
        onChangeText={setQuantity}
      />
      <TouchableOpacity style={styles.saveText} onPress={saveGroceryItem}>
        <Text>Save</Text>
      </TouchableOpacity>
    </View>
  );
};
