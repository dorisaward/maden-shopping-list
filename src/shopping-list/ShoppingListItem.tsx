import React from 'react';
import {ListRenderItem, Text, View} from 'react-native';
import {GroceryItem} from '../groceries/GroceryItem';
import {styles} from '../styles';

export const ShoppingListItem: ListRenderItem<GroceryItem> = ({item}) => (
  <View style={styles.shoppingListItem}>
    <Text style={styles.shoppingListItemText}>{item.name}</Text>
    <Text style={styles.shoppingListItemText}>{item.quantity}</Text>
  </View>
);
