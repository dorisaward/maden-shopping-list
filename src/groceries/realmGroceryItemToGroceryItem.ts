import {GroceryItem} from './GroceryItem';
import {RealmGroceryItem} from './RealmGroceryItem';

export const realmGroceryItemToGroceryItem = (
  realmGroceryItem: RealmGroceryItem,
): GroceryItem => ({
  _id: realmGroceryItem._id,
  name: realmGroceryItem.name,
  quantity: realmGroceryItem.quantity,
});
