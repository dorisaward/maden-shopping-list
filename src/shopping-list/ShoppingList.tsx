import React, { useState } from 'react';
import {FlatList, FlatListProps} from 'react-native';
import {ShoppingListItem} from './ShoppingListItem';
import {GroceryItem} from '../groceries/GroceryItem';

type Props = {
  initialGroceryItems: GroceryItem[];
};

export const ShoppingList = ({
  initialGroceryItems,
}: Props): React.JSX.Element => {
const [groceryItems, setGroceryItems] = useState<GroceryItem[]>(initialGroceryItems)

  return (
    <FlatList
      data={groceryItems}
      renderItem={ShoppingListItem}
      keyExtractor={keyExtractor}
    />
  );
};

const keyExtractor: FlatListProps<GroceryItem>['keyExtractor'] = item =>
  item.id;
