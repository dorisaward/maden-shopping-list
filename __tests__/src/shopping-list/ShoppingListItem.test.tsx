import 'react-native';
import React from 'react';
import {it, expect} from '@jest/globals';
import renderer from 'react-test-renderer';
import {ShoppingListItem} from '../../../src/shopping-list/ShoppingListItem';
import {GroceryItem} from '../../../src/groceries/GroceryItem';
import {ListRenderItemInfo} from 'react-native';

it('should render shopping list item', () => {
  // Given
  const item: GroceryItem = {
    id: 'test-id',
    name: 'test name',
    quantity: 100,
  };
  const renderable = <ShoppingListItem {...otherProps} item={item} index={0} />;

  // When
  const {toJSON} = renderer.create(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});

const otherProps: Pick<ListRenderItemInfo<GroceryItem>, 'separators'> = {
  separators: {
    highlight: jest.fn(),
    unhighlight: jest.fn(),
    updateProps: jest.fn(),
  },
};
