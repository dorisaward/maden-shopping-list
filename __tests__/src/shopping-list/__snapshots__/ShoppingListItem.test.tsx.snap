// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render shopping list item 1`] = `
<View
  style={
    {
      "borderColor": "#6495ed",
      "borderRadius": 30,
      "borderWidth": 3,
      "flexDirection": "row",
      "marginBottom": 3,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 15,
        "padding": 10,
      }
    }
  >
    test name
  </Text>
  <Text
    style={
      {
        "fontSize": 15,
        "padding": 10,
      }
    }
  >
    100
  </Text>
</View>
`;
